app/src/main/java/com/fantastic/bookxchange/activities/NearMeActivity.java

```diff
@@ -0,0 +1,244 @@
+
+    
+    //Data
+    
+    List<User> users;
+    HashMap<Book, List<Marker>> books;
+    
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_near_me);
+
+        if (savedInstanceState != null && savedInstanceState.keySet().contains(KEY_LOCATION)) {
+            // Since KEY_LOCATION was found in the Bundle, we can be sure that mCurrentLocation
+            // is not null.
+            mCurrentLocation = savedInstanceState.getParcelable(KEY_LOCATION);
+        }
+        setupMap();
+
+    }
+
+    private void populateData() {
+        users = DataTest.fakeData();
+
+        books = new HashMap<>();
+
+        for (User u : users) {
+            if (!u.getBooks().isEmpty()){
+                Marker marker = MapUtils.addMarker(map, u);
+                for (Book b : u.getBooks()) {
+                    if (books.containsKey(b)) {
+                        books.get(b).add(marker);
+                    } else {
+                        List<Marker> markersList = new ArrayList<>();
+                        markersList.add(marker);
+                        books.put(b, markersList);
+                    }
+                }
+            }
+        }
+
+        //TODO Show book data on fragment
+    }
+
+    private void setupMap() {
+        if (TextUtils.isEmpty(getResources().getString(R.string.google_maps_api_key))) {
+            throw new IllegalStateException("You forgot to supply a Google Maps API key");
+        }
+        mapFragment = ((SupportMapFragment) getSupportFragmentManager().findFragmentById(R.id.map));
+        if (mapFragment != null) {
+            mapFragment.getMapAsync(map1 -> loadMap(map1));
+        } else {
+            toast("Error - Map Fragment was null!!");
+        }
+
+    }
+
+    protected void loadMap(GoogleMap googleMap) {
+        map = googleMap;
+        if (map != null) {
+            NearMeActivityPermissionsDispatcher.getMyLocationWithCheck(this);
+            NearMeActivityPermissionsDispatcher.startLocationUpdatesWithCheck(this);
+            map.setOnMarkerClickListener(marker -> {
+                User u = (User)marker.getTag();
+
+                Geocoder geocoder;
+                List<Address> addresses;
+                geocoder = new Geocoder(NearMeActivity.this, Locale.getDefault());
+
+                try {
+                    addresses = geocoder.getFromLocation(u.getLocation().latitude, u.getLocation().longitude, 1);
+                    Log.d("NearMe", u.getName() + " location: " + addresses.get(0).getLocality() + ", " + addresses.get(0).getAdminArea());
+                } catch (IOException e) {
+                    e.printStackTrace();
+                }
+                //TODO Intent to User Profile
+
+                return true;
+            });
+            populateData();
+        } else {
+            toast("Error - Map was null!!");
```

Comment:
think better to put this string in resource and use `toast(R.string.name_of_str);`